### ISSUE TITLE
Indexing debug performance

### ISSUE THREAD

**Issue by psiha** [jfalcou/kiwaku#64]
If you remember this is something I fixed in the previous version (before the last major PR) - we have a serious issue with indexing performance in builds that have asserts and sanitizers enabled, even if all other optimizations are turned on (the attached image shows this in the case of a reference convolution implementation - that does the triple loop and indexing - and this is with inline and flatten attributes slapped all over the place) - it actually causes unit tests on xeon servers to timeout.
This is also connected to:
- https://github.com/jfalcou/kiwaku/issues/39 (necessity to do the stride reversing dance)
- https://github.com/jfalcou/kiwaku/issues/63

![image](https://user-images.githubusercontent.com/340735/207101371-602228f0-179e-4c76-a7be-e97cdf2eb1f0.png)

**Comment by psiha** [comment]
doesn't help much (for some reason clang does not even flatten it all fully, at least with sanitizers enabled), here is codegen for kumi make_tuple (or something related :D)
`_ZN4kumi9push_backINS_5tupleIJiEEEiEEDaRKT_OT0_:
0x0000555556e2abd0  jmp    0x555556e2abd8 <_ZN4kumi9push_backINS_5tupleIJiEEEiEEDaRKT_OT0_+8> 
0x0000555556e2abd2  jbe    0x555556e2ac06 <_ZN4kumi9push_backINS_5tupleIJiEEEiEEDaRKT_OT0_+54> 
0x0000555556e2abd4  enter  $0xd747,$0x0 
0x0000555556e2abd8  push   %rbx 
0x0000555556e2abd9  mov    %rsi,%rbx 
0x0000555556e2abdc  call   0x555556e2ac40 <_ZN4kumi3getILm0EJiEEEDcRKNS_5tupleIJDpT0_EEE> 
0x0000555556e2abe1  mov    %rax,%rcx 
0x0000555556e2abe4  shr    $0x3,%rcx 
0x0000555556e2abe8  movzbl 0x7fff8000(%rcx),%ecx 
0x0000555556e2abef  test   %cl,%cl 
0x0000555556e2abf1  jne    0x555556e2ac12 <_ZN4kumi9push_backINS_5tupleIJiEEEiEEDaRKT_OT0_+66> 
0x0000555556e2abf3  mov    (%rax),%ecx 
0x0000555556e2abf5  mov    %rbx,%rax 
0x0000555556e2abf8  shr    $0x3,%rax 
0x0000555556e2abfc  movzbl 0x7fff8000(%rax),%eax 
0x0000555556e2ac03  test   %al,%al 
0x0000555556e2ac05  jne    0x555556e2ac26 <_ZN4kumi9push_backINS_5tupleIJiEEEiEEDaRKT_OT0_+86> 
0x0000555556e2ac07  mov    (%rbx),%eax 
0x0000555556e2ac09  shl    $0x20,%rax 
0x0000555556e2ac0d  or     %rcx,%rax 
0x0000555556e2ac10  pop    %rbx 
0x0000555556e2ac11  ret     
0x0000555556e2ac12  mov    %eax,%edx 
0x0000555556e2ac14  and    $0x7,%dl 
0x0000555556e2ac17  add    $0x3,%dl 
0x0000555556e2ac1a  cmp    %cl,%dl 
0x0000555556e2ac1c  jl     0x555556e2abf3 <_ZN4kumi9push_backINS_5tupleIJiEEEiEEDaRKT_OT0_+35> 
0x0000555556e2ac1e  mov    %rax,%rdi 
0x0000555556e2ac21  call   0x555556b9b9e0 <__asan_report_load4> 
0x0000555556e2ac26  mov    %ebx,%edx 
0x0000555556e2ac28  and    $0x7,%dl 
0x0000555556e2ac2b  add    $0x3,%dl 
0x0000555556e2ac2e  cmp    %al,%dl 
0x0000555556e2ac30  jl     0x555556e2ac07 <_ZN4kumi9push_backINS_5tupleIJiEEEiEEDaRKT_OT0_+55> 
0x0000555556e2ac32  mov    %rbx,%rdi 
0x0000555556e2ac35  call   0x555556b9b9e0 <__asan_report_load4> `

and the stacktrace:
![image](https://user-images.githubusercontent.com/340735/208998169-43a99273-a012-42af-9295-b89012d90aad.png)

so you need to test/profile this (indexing multidimensional arrays in nested loops) w/ sanitizers enabled...(i'd start with pass-by-value wherever possible:)

**Comment by jfalcou** [comment]
Yeah, so that's kumi that need some flatten/force inlining massage.
You won't be surprised I ussually don't check those scenarios :/

**Comment by jfalcou** [comment]
OK, culprit is kum fold algorithm that for some reason are really bad in some situation.
Gonna wrtie a custom one for Kiwaku and fix Kumi later

**Comment by psiha** [comment]
ok, finally back to kiwaku 😀 

so, on this topic - this is how i solved it in the mean time:
`    [[ using gnu: pure, flatten ]]
    size_t index( auto const ... coordinates ) const noexcept
    requires( sizeof...( coordinates ) == cardinality )
    {
        auto const extents{ as_array() };
#   if !defined( NDEBUG ) // kwk index() does not verify its inputs
        [ = ]< int... i >( std::integer_sequence< int, i... > ) __attribute__(( always_inline ))
        {
            MB_CONSTEXPR_ASSERT(( ( coordinates < extents[ i ] ) && ... ));
        }( IntegerSequence{} );
#   endif
        // https://github.com/jfalcou/kiwaku/issues/64
        //return safeCast< size_t >( as_stride().linearize( coordinates... ) );
        size_t strides[ cardinality ];
        strides[ 0 ] = 1;
        [ & ]< int... i >( std::integer_sequence< int, i... > ) __attribute__(( always_inline ))
        {
            ( ..., (strides[ i + 1 ] = strides[ i ] * extents[ cardinality - 1 - i ]) );
        }( std::make_integer_sequence< int, cardinality - 1 >{} );

        return [ = ]< int... i >( std::integer_sequence< int, i... > ) __attribute__(( always_inline ))
        {
            return ( ( strides[ cardinality - 1 - i ] * coordinates ) + ... );
        }( IntegerSequence{} );
    }`

**Comment by psiha** [comment]
now off to go through rest of your changes and questions...and figuring out what i have to fix after the update to latest master :D
![image](https://user-images.githubusercontent.com/340735/222702614-350a6e2f-67aa-438f-97a5-dc62f82c2084.png)
